### ISSUE TITLE
False positive on only_throw_errors lint

### ISSUE THREAD

**Issue by jonahwilliams** [dart-lang/linter#2637]
The following code leads to an `only_throw_errors` diagnostic, but `foo` is known to be an Exception at this point. Casting to exception leads to an unnecessary cast lint. This does not repro if the static type of `foo` is changed to Object or dynamic.

```dart
abstract class Foo {}

void bar(Foo foo) {
  if (foo is Exception) {
    throw foo;
  }
}
```

@pq

**Comment by bwilkerson** [comment]
I believe that the reason for the observed behavior is that the `foo is Exception` doesn't promote `foo` because there is no relationship between `Foo` and `Exception`. The reason is works in the second case is because there _is_ a relationship between `Foo` and `FooException`.

I'm not sure that we can "fix" the lint, given that the type system can't tell us that `foo` is an `Exception`.

**Comment by jonahwilliams** [comment]
Then maybe the problem is the cast diagnostic instead? This doesn't repro on my example but in the issue I linked I also got a diagnostic for an unnecessary cast error in flutter_reactive_ble - so that example might be different.

**Comment by jonahwilliams** [comment]
i.e. might be a bad reduced example

**Comment by bwilkerson** [comment]
Ah, then yes, this is a bug in the lint. When we have a `TypeParameterType` we need to check the `promotedBound` as well as the original type.

**Comment by srawlins** [comment]
Given that `dynamic` is allowed by the lint, I think an unbound type parameter should also be allowed:

```dart
f(dynamic e) => throw e; // OK
g<T>(T e) => throw e; // OK
```
